lib/dao.js

```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
 };
 
 /**
+ * Replace or insert a model instance: replace exiting record if one is found, such that parameter `data.id` matches `id` of model instance;
+ * otherwise, insert a new record.
+ *
+ * NOTE: No setters, validations, or hooks are applied when using upsert.
```

Comment:
This is not correct, we should certainly validate the instance in `replaceOrCreate`, since it is always a full instance (instead of a partial update in case of `updateOrCreate`). I think we should apply setters and hooks too.

@raymondfeng thoughts?